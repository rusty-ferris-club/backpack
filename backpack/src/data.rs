use anyhow::Result as AnyResult;
use core::fmt::Debug;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Clone, Debug)]
pub struct Location {
    pub url: String,
    pub domain: String,
    pub path: String,
    pub project: String,
    pub port: Option<u16>,
    pub query: Option<String>,
    pub subfolder: Option<String>,
    pub gref: Option<String>,
    pub username: String,
    pub password: Option<String>,
    pub is_git: bool,
}
impl Location {
    pub fn git_url(&self) -> String {
        format!(
            "git@{}:{}.git",
            self.domain,
            self.path.trim_start_matches('/')
        ) // needs: auth, port
    }

    pub fn web_url(&self) -> String {
        format!("https://{}{}", self.domain, self.path) // needs: auth, port
    }

    /// Create a logical `Location` from a `Url`.
    ///
    /// # Errors
    ///
    /// This function will return an error if `url` is illegal
    pub fn from(url: &Url, is_git: bool) -> AnyResult<Self> {
        let path = url.path();
        let parts = path.split("/-/").collect::<Vec<_>>();
        let (path, subfolder) = if parts.len() == 2 {
            (parts[0], Some(parts[1].to_string()))
        } else {
            (path, None)
        };

        let project = path
            .split('/')
            .last()
            .ok_or_else(|| anyhow::anyhow!("cannot find project name"))?;

        Ok(Self {
            url: url.to_string(),
            domain: url.domain().unwrap_or_default().to_string(),
            path: path.trim_end_matches(".git").to_string(),
            project: project.to_string(),
            port: url.port(),
            query: url.query().map(str::to_string),
            subfolder,
            gref: url.fragment().map(str::to_string),
            username: url.username().to_string(),
            password: url.password().map(str::to_string),
            is_git,
        })
    }
}

#[derive(Debug)]
pub struct Archive {
    pub url: String,
    pub root: ArchiveRoot,
}

#[derive(Debug)]
pub enum ArchiveRoot {
    Folder(String),
    FirstFolder,
    None,
}
#[derive(Debug)]
pub struct Assets {
    pub archive: Option<Archive>,
    pub git: Option<String>,
}

pub enum Overwrite {
    Ask,
    Always,
    Never,
    Custom(Box<dyn Fn(&str) -> bool>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum CopyMode {
    #[serde(rename = "new")]
    Copy,
    #[serde(rename = "apply")]
    Apply,
    #[serde(rename = "all")]
    All,
}

#[allow(clippy::struct_excessive_bools)]
#[derive(Debug)]
pub struct Opts {
    pub show_progress: bool,
    pub overwrite: bool,
    pub mode: CopyMode,
    pub is_git: bool,
    pub no_cache: bool,
    pub always_yes: bool,
    pub remote: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_location_urls() {
        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            true,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc =
            Location::from(&Url::parse("https://github.com/user/repo").unwrap(), false).unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            false,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");
    }
}
