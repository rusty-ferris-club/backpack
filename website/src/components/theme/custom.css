/**
 * Any CSS included here will be global. The classic template
 * bundles Infima by default. Infima is a CSS framework designed to
 * work well for content-centric websites.
 */

/* You can override the default Infima variables here. */
@import url('/static/fonts/inter/inter.css');

body {
  font-feature-settings: "kern", "cv02", "cv03", "cv04", "cv11";
}

.main-wrapper {
  max-width: var(--chakra-sizes-container-xl2);
  margin: 0 auto;
  width: 100%;
}

.navbar {
  padding: 0;
}

.navbar__inner {
  width: 100%;
  margin-inline-start: auto;
  margin-inline-end: auto;
  max-width: var(--chakra-sizes-container-lg);
}

.navbar__logo {
  height: 48px;
}

.docs-wrapper body {
  background: var(--chakra-colors-t_background_docs);
}

.docs-wrapper .navbar {
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.05) !important;
  transition: all 300ms;
}

.docs-wrapper .navbar__inner {
  max-width: var(--chakra-sizes-container-xl2);
}

.footer__logo {
  max-width: inherit;
}

.docs-wrapper .main-wrapper {
  background-color: var(--chakra-colors-t_background_docs)
}

.menu__list-item-collapsible--active,
.menu__link--active:not(.menu__link--sublist),
.menu__link:hover,
.menu__list-item-collapsible:hover {
  background: none;
  color: inherit;
}

:root {
  --ifm-color-primary: #F545A1;
  --ifm-color-primary-dark: #F545A1;
  --ifm-color-primary-darker: #F545A1;
  --ifm-color-primary-darkest: #F545A1;
  --ifm-color-primary-light: #F545A1;
  --ifm-color-primary-lighter: #F545A1;
  --ifm-color-primary-lightest: #F545A1;
  --ifm-background-surface-color: #fff;
  --ifm-code-font-size: 95%;
  --ifm-font-family-base: 'Inter var';
  --ifm-heading-font-family: 'Jost';
  --ifm-toc-border-color: rgba(0, 0, 0, 0.05);
  --docusaurus-highlighted-code-line-bg: rgba(0, 0, 0, 0.1);
  --ifm-navbar-search-input-background-color: #F2F3F1;
  --ifm-menu-link-sublist-icon: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="16px" height="16px" viewBox="0 0 24 24"><path fill="rgba(0,0,0)" d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"></path></svg>');
  --ifm-font-color-base: var(--chakra-colors-t_text);
}

/* For readability concerns, you should choose a lighter palette in dark mode. */
[data-theme='dark'] {
  --ifm-color-primary: #F545A1;
  --ifm-color-primary-dark: #F545A1;
  --ifm-color-primary-darker: #F545A1;
  --ifm-color-primary-darkest: #F545A1;
  --ifm-color-primary-light: #F545A1;
  --ifm-color-primary-lighter: #F545A1;
  --ifm-color-primary-lightest: #F545A1;
  --ifm-navbar-search-input-background-color: #444;
  --docusaurus-highlighted-code-line-bg: #484848;
  --docusaurus-highlighted-code-line-bg: rgba(0, 0, 0, 0.3);
  --ifm-font-color-base: var(--chakra-colors-canvas);
}

.alert {
  border-left-width: 0;
  box-shadow: none;
}

.theme-doc-breadcrumbs {
  display: none;
}

.docs-wrapper main .container {
  padding: 0 !important;
}

.col.col--3 {
  border-left: 1px solid var(--ifm-toc-border-color);
}

.table-of-contents {
  font-size: 14px;
}

.table-of-contents__left-border {
  border: 0;
}

.table-of-contents__link,
.table-of-contents__link:hover {
  color: var(--chakra-colors-t_text_weakest);
}

.table-of-contents__link--active {
  color: var(--chakra-colors-t_text);
  font-weight: 600;
}

.table-of-contents__link--active:hover {
  color: var(--chakra-colors-t_text);
}

.docs-wrapper article {
  padding: 0 3em;
}

.docs-wrapper article a {
  color: var(--ifm-link-color);
  text-decoration: var(--ifm-link-decoration);
}

.docs-wrapper article img {
  border-radius: 6px;
  box-shadow: 0 4px 15px -1px rgb(0 0 0 / 40%);
}

.pagination-nav {
  padding: 1em 2em 2em 2em;
}

.docs-wrapper .navbar {
  background-color: var(--chakra-colors-t_background_docs);
}

.menu {
  font-weight: 400;
  font-size: 14px;
}

.menu__link--active,
.menu__link--active:hover {
  color: var(--chakra-colors-t_text);
  font-weight: 600;
}

.menu__list {
  padding: 1em;
}


.menu__list .menu__list {
  padding: 0;
  margin-left: 13px;
}

.menu__list-item .menu__list {
  border-left: 1px solid var(--chakra-colors-t_text);
  ;
}

.menu__list-item:not(:first-child) {
  margin-top: 0.5em;
}

.menu__link {
  padding: var(--ifm-menu-link-padding-vertical) 5px;
  color: var(--chakra-colors-t_text_weakest);
}

.menu__caret {
  padding: 0 4px;
}

.menu__caret:before {
  color: var(--chakra-colors-t_text);
  transform: rotate(180deg) scale(0.7);
}

.menu__list-item--collapsed .menu__link--sublist:after,
.menu__list-item--collapsed .menu__caret:before {
  transform: rotate(90deg) scale(0.7);
}

.markdown h1:first-child,
.markdown h1 {
  margin: 2em 0 1em 0;
  font-weight: 500;
  font-size: 3em;
  color: var(--chakra-colors-t_text_docs);
}

.markdown>h2 {
  font-size: 2.2em;
  font-weight: 500;
  margin: 2em 0 1em 0;
  color: var(--chakra-colors-t_text_docs);
}

.markdown>h3 {
  font-size: 1.7em;
  font-weight: 500;
  margin: 2em 0 1em 0;
  color: var(--chakra-colors-t_text_docs);
}

.markdown>h4 {
  font-weight: 600;
  margin: 2em 0 1em 0;
  color: var(--chakra-colors-t_text_docs);
}

code {
  border: 0;
  vertical-align: inherit;
}
