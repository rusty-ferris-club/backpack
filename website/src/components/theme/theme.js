import { extendTheme } from '@chakra-ui/react'
import { mode } from '@chakra-ui/theme-tools'

// 2. Call `extendTheme` and pass your custom values
const theme = extendTheme({
  fonts: {
    heading: "'Jost', sans-serif",
    body: "'Inter var', sans-serif",
  },
  styles: {
    global: {
      body: {
        fontFeatureSettings: '"cv02", "cv03", "cv04", "cv11"',
        background: 't_background',
        color: 't_text',
      },
      h1: { letterSpacing: '-0.03em', color: 't_text' },
      h2: { letterSpacing: '-0.03em', color: 't_text' },
      h3: { letterSpacing: '-0.02em', color: 't_text' },
      svg: {
        display: 'inline',
      },
    },
  },
  sizes: {
    max: '100%',
    container: {
      xl2: '1440px',
    },
  },
  colors: {
    brand: {
      50: '#FFF0F0',
      100: '#FBB1D8',
      200: '#FA99CC',
      300: '#F87CBD',
      400: '#F655A9',
      500: '#F545A1',
      600: '#F3208E',
      700: '#DF0C7A',
      800: '#C20A6A',
      900: '#A5095A',
    },
    charcoal: '#111',
    canvas: '#ffffff',
    outline: 'blue',
    text: '#444',
    textLighter: '#555',
    textLightest: '#666',
    textDarker: '#222',
    textDocs: '#555',
  },
  semanticTokens: {
    colors: {
      t_border_color: {
        default: 'blackAlpha.50',
        _dark: 'whiteAlpha.300',
      },
      t_strong: {
        default: 'textDarker',
        _dark: 'whiteAlpha.900',
      },
      t_text_docs: {
        default: 'textDocs',
        _dark: 'whiteAlpha.800',
      },
      t_text: {
        default: 'text',
        _dark: 'whiteAlpha.800',
      },
      t_weak: {
        default: 'textLighter',
        _dark: 'whiteAlpha.800',
      },
      t_weakest: {
        default: 'textLightest',
        _dark: 'whiteAlpha.800',
      },
      t_background: {
        default: 'canvas',
        _dark: 'transparent',
      },
      t_background_docs: {
        default: 'canvas',
        _dark: 'charcoal',
      },
      t_background_article: {
        default: 'white',
        _dark: 'charcoal',
      },
      ink: {
        default: 'text',
        _dark: 'canvas',
      },
      inverseInk: {
        default: 'canvas',
        _dark: 'text',
      },
    },
  },
  shadows: {
    outline: '0 0 0 3px #FB309Aca',
  },
  components: {
    Link: {
      baseStyle: {
        color: 'brand.500',
        transition: 'color 200ms',
        _hover: {
          color: 'brand.500',
        },
        _focus: {
          boxShadow: 'none',
        },
      },
    },
    Code: {
      variants: {
        installer: (props) => ({
          border: 'none',
          background: 'none',
          color: mode('black', 'white')(props),
          fontSize: 16,
        }),
      },
    },
    Button: {
      sizes: {
        xl: {
          h: '60px',
          minW: 16,
          fontSize: 'md',
          px: 7,
        },
      },
      variants: {
        outline: {
          borderWidth: '2px',
          _hover: {
            textDecoration: 'none',
          },
        },
        'clipboard-copy': (props) => ({
          bg: 'ink',
          color: 'inverseInk',
          _focus: {
            shadow: 'none',
          },
          _hover: {
            bg: mode('blackAlpha.700', 'whiteAlpha.900')(props),
          },
          _active: {
            bg: mode('blackAlpha.800', 'whiteAlpha.800')(props),
          },
        }),
      },
    },
  },
})
console.log('theme', theme)
export default theme
