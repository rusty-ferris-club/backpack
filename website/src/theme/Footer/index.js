// docusaurus(J): changed footer completely
import React from 'react'
import { useThemeConfig } from '@docusaurus/theme-common'
import FooterLogo from '@theme/Footer/Logo'
import FooterCopyright from '@theme/Footer/Copyright'
import {
  Box,
  Link as CLink,
  Container,
  Flex,
  HStack,
  Text,
  VStack,
} from '@chakra-ui/react'
import Link from '@docusaurus/Link'

function Footer() {
  const { footer } = useThemeConfig()

  if (!footer) {
    return null
  }

  const { copyright, links, logo } = footer
  return (
    <Box borderTop="1px solid" borderColor="t_border_color">
      <Container maxW="container.lg">
        <Flex sx={{ my: 10 }} flexDirection="column" alignItems="center">
          <Flex
            w="100%"
            flexDirection={['column', 'column', 'row']}
            alignItems={['center', 'center', 'flex-start']}
          >
            <Flex flex={1}>
              <VStack sx={{ mb: 6 }}>
                <Box sx={{ maxW: '100px' }}>
                  <FooterLogo logo={logo} />
                </Box>
              </VStack>
            </Flex>
            <HStack alignItems="flex-start" spacing="64px">
              {links &&
                links.length > 0 &&
                links.map(({ title, items }) => (
                  <VStack alignItems="flex-start" key={title}>
                    <Text fontWeight={600}>{title}</Text>
                    <VStack alignItems="flex-start">
                      {items.map(({ label, href, to }) => (
                        <CLink
                          as={Link}
                          sx={{ color: 'inherit' }}
                          key={href || to}
                          to={to}
                          href={href}
                        >
                          {label}
                        </CLink>
                      ))}
                    </VStack>
                  </VStack>
                ))}
            </HStack>
          </Flex>
          <Box sx={{ mt: 14 }}>
            {copyright && <FooterCopyright copyright={copyright} />}
          </Box>
        </Flex>
      </Container>
    </Box>
  )
}

export default React.memo(Footer)
