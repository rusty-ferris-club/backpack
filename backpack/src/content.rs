use crate::data::CopyMode;
use crate::data::Location;
use crate::data::Overwrite;
use anyhow::Result;
use dialoguer;
use dialoguer::theme::ColorfulTheme;
use std::fs;
use std::path::{Path, PathBuf};
use tracing::warn;
use walkdir;

#[derive(Default)]
pub struct Deployer {}

#[tracing::instrument(skip_all, err)]
fn copy(source: &Path, dest: &Path, is_file: bool, overwrite: Overwrite) -> Result<()> {
    if is_file {
        // dest is a full path incl. file
        let dest_path = dest
            .parent()
            .ok_or_else(|| anyhow::anyhow!("cannot get parent for {:?}", dest))?;
        if !dest_path.exists() {
            println!("{:?}", dest_path);
            fs::create_dir_all(&dest_path)?;
        }

        println!("copy {:?}, {:?}", source, dest);
        fs::copy(source, &dest)?;
        return Ok(());
    }
    walkdir::WalkDir::new(source)
        .into_iter()
        .try_for_each(|entry| {
            let entry = entry?;
            let path = entry.path();
            if path.is_file() {
                if let Some(parent) = path.parent() {
                    let dest_parent = dest.join(parent.strip_prefix(source)?);
                    if !dest_parent.exists() {
                        // Create the same dir concurrently is ok according to the docs.
                        fs::create_dir_all(dest_parent)?;
                    }
                }
                let to = dest.join(path.strip_prefix(source)?);
                if to.exists() {
                    let should_copy = match overwrite {
                        Overwrite::Always => true,
                        Overwrite::Ask => {
                            prompt(&format!("'{}' already exists. overwrite?", to.display()))
                        }
                        _ => false,
                    };
                    if should_copy {
                        fs::copy(path, to)?;
                    }
                } else {
                    fs::copy(path, to)?;
                }
            }

            anyhow::Ok(())
        })?;
    Ok(())
}

impl Deployer {
    #[tracing::instrument(skip_all, err)]
    pub fn deploy(
        &self,
        source: &Path,
        dest: Option<&Path>,
        location: &Location,
        mode: &CopyMode,
        overwrite: bool,
        remove_source: bool,
    ) -> Result<String> {
        // xxx: either way canonicalize paths.
        let final_source = source.join(location.subfolder.clone().unwrap_or_default());
        let dest = dest.map(std::path::Path::to_path_buf);
        let location_path = location.subfolder.clone().map(PathBuf::from);

        // is this "deploying" a single file or a folder?
        let is_file = final_source.is_file();
        let final_dest = if is_file {
            // final dest = dest | location+fname | fname
            let fname = final_source
                .file_name()
                .ok_or_else(|| anyhow::anyhow!("cannot get file name for {:?}", final_source))?;
            dest.or_else(|| {
                location_path.and_then(|loc| loc.parent().map(|p| p.to_path_buf().join(fname)))
            })
            .unwrap_or_else(|| PathBuf::from(fname))
        } else {
            dest.or(location_path)
                .unwrap_or_else(|| PathBuf::from(".".to_string()))
        };

        match mode {
            CopyMode::Copy => {
                if final_dest.exists() {
                    anyhow::bail!("path already exists: {}", final_dest.display());
                }
                copy(&final_source, &final_dest, is_file, Overwrite::Always)?;
            }
            CopyMode::Apply => {
                copy(
                    &final_source,
                    &final_dest,
                    is_file,
                    if overwrite {
                        Overwrite::Always
                    } else {
                        Overwrite::Ask
                    },
                )?;
            }
            CopyMode::All => {}
        }
        if remove_source {
            // xxx don't remove for now
            warn!("remove requested, but not removing '{}'", source.display());
        }
        // copy vs apply
        Ok(final_dest.display().to_string())
    }
}

fn prompt(q: &str) -> bool {
    dialoguer::Confirm::with_theme(&ColorfulTheme::default())
        .with_prompt(q)
        .interact()
        .unwrap()
}
