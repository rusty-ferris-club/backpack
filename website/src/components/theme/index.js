/* eslint-disable react/prop-types */
import React from 'react'
import { ChakraProvider } from '@chakra-ui/react'
import theme from './theme'
import ResetCSS from './reset-css'

export const Theme = ({ children }) => {
  return (
    <ChakraProvider theme={theme} resetCSS={false}>
      <ResetCSS />
      {children}
    </ChakraProvider>
  )
}
