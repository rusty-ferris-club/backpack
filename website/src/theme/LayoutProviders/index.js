// docusaurus(J): add sync color mode to sync colors
import React from 'react'
import LayoutProviders from '@theme-original/LayoutProviders'
import { Theme } from '../../components/theme'
import { SyncColorMode } from '../../utils/useSyncColorMode'

export default function LayoutProvidersWrapper({ children, ...props }) {
  return (
    <>
      <LayoutProviders {...props}>
        <Theme>
          <SyncColorMode />
          {children}
        </Theme>
      </LayoutProviders>
    </>
  )
}
