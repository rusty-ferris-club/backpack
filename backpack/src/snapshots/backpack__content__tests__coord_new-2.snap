---
source: backpack/src/content.rs
expression: "Coordinate::new(Path::new(\"here\"), None,\n        &Location::from(&Url::parse(\"https://github.com/foo/bar/-/subfolder/qux\").unwrap(),\n                    true).unwrap(), false).unwrap()"
---
from: here/subfolder/qux
to: subfolder/qux
is_file: false
remove_from: false

